app/src/crypto.c

```diff
@@ -584,3 +612,288 @@ zxerr_t crypto_fillMASP(uint8_t *buffer, uint16_t bufferLen, uint16_t *cmdRespon
 
     return zxerr_ok;
 }
+
+static parser_error_t h_star(uint8_t *a, uint16_t a_len, uint8_t *b, uint16_t b_len, uint8_t *output) {
+    if (a == NULL || b == NULL || output == NULL) {
+        return parser_no_data;
+    }
+
+    uint8_t hash[BLAKE2B_OUTPUT_LEN] = {0};
+#if defined(LEDGER_SPECIFIC)
+    cx_blake2b_t ctx = {0};
+    ASSERT_CX_OK(cx_blake2b_init2_no_throw(&ctx, BLAKE2B_OUTPUT_LEN, NULL, 0, (uint8_t *)SINGNING_REGJUBJUB,
+                                           sizeof(SINGNING_REGJUBJUB)));
+    ASSERT_CX_OK(cx_blake2b_update(&ctx, a, a_len));
+    ASSERT_CX_OK(cx_blake2b_update(&ctx, b, b_len));
+    cx_blake2b_final(&ctx, hash);
+#else
+    blake2b_state state = {0};
+    blake2b_init_with_personalization(&state, BLAKE2B_OUTPUT_LEN, (const uint8_t *)SINGNING_REGJUBJUB,
+                                      sizeof(SINGNING_REGJUBJUB));
+    blake2b_update(&state, a, a_len);
+    blake2b_update(&state, b, b_len);
+    blake2b_final(&state, hash, BLAKE2B_OUTPUT_LEN);
+#endif
+
+    from_bytes_wide(hash, output);
+
+    return parser_ok;
+}
+static zxerr_t sign_sapling_spend(keys_t *keys, uint8_t alpha[static KEY_LENGTH], uint8_t sign_hash[static KEY_LENGTH], uint8_t *signature) {
+    if (alpha == NULL || sign_hash == NULL || signature == NULL) {
+        return zxerr_no_data;
+    }
+
+    uint8_t data_to_be_signed[2 * HASH_LEN] = {0};
+    uint8_t rsk[KEY_LENGTH] = {0};
+    uint8_t rk[KEY_LENGTH] = {0};
+
+    // get randomized secret
+    randomized_secret_from_seed(keys->ask, alpha, rsk);
+
+    //rsk to rk
+    scalar_multiplication(rsk, SpendingKeyGenerator, rk);
+
+    // sign
+    MEMCPY(data_to_be_signed, rk, KEY_LENGTH);
+    MEMCPY(data_to_be_signed + KEY_LENGTH, sign_hash, HASH_LEN);
+
+    // Get rng
+    uint8_t rng[RNG_LEN] = {0};
+    cx_rng_no_throw(rng, RNG_LEN);
+
+    // Compute r and rbar
+    uint8_t r[32] = {0};
+    uint8_t rbar[32] = {0};
+    CHECK_PARSER_OK(h_star(rng, sizeof(rng), data_to_be_signed, sizeof(data_to_be_signed), r));
+    CHECK_PARSER_OK(scalar_multiplication(r, SpendingKeyGenerator, rbar));
+
+    //compute s and sbar
+    uint8_t s[32] = {0};
+    uint8_t sbar[32] = {0};
+    CHECK_PARSER_OK(h_star(rbar, sizeof(rbar), data_to_be_signed, sizeof(data_to_be_signed), s));
+    CHECK_PARSER_OK(compute_sbar(s, r, rsk, sbar));
+
+    MEMCPY(signature, rbar, HASH_LEN);
+    MEMCPY(signature + HASH_LEN, sbar, HASH_LEN);
+
+    return zxerr_ok;
+}
+
+zxerr_t crypto_sign_spends_sapling(const parser_tx_t *txObj, keys_t *keys, uint8_t *output, uint16_t outputLen, uint16_t *responseLen) {
+    zemu_log_stack("crypto_signspends_sapling");
+    if (txObj->transaction.sections.maspTx.data.sapling_bundle.n_shielded_spends == 0) {
+        return zxerr_ok;
+    }
+
+    MEMZERO(output, outputLen);
```

Comment:
Add a check for `outputLen >= txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends * MASP_SIG_LEN`